src/plots/cartesian/set_convert.js

```diff
@@ -371,49 +381,99 @@ module.exports = function setConvert(ax, fullLayout) {
-                            var v0 = arrayIn[0][j];
-                            var v1 = arrayIn[1][j];
-
-                            if(isValidCategory(v0) && isValidCategory(v1)) {
-                                list.push([v0, v1]);
-
-                                if(!(v0 in seen[0][1])) {
-                                    seen[0][1][v0] = seen[0][0]++;
-                                }
-                                if(!(v1 in seen[1][1])) {
-                                    seen[1][1][v1] = seen[1][0]++;
-                                }
+                    if(isArrayOrTypedArray(arrayIn[0])) {
+                        var arrays = arrayIn.map(function(x) {
+                            return x;
+                        });
+                        var valLetter;
+                        if(trace.type === 'ohlc' | trace.type === 'candlestick') {
+                            var t = trace;
+                            var valsTransform = sortLib.transpose([t.open, t.high, t.low, t.close]);
+                            arrays.push(valsTransform);
+                        } else if(trace.z) {
+                            if(axLetter === 'x') {
+                                arrays.push(sortLib.transpose(trace.z));
+                            } else {
+                                arrays.push(trace.z);
                             }
+                            valLetter = 'z';
+                        } else if(axLetter === 'y' && trace.x) {
+                            arrays.push(trace.x);
+                            valLetter = 'x';
+                        } else if(trace.y) {
+                            arrays.push(trace.y);
+                            valLetter = 'y';
+                        } else {
+                            var nullArray = arrayIn[0].map(function() {return null;});
+                            arrays.push(nullArray);
+                        }
+                        var objList = sortLib.matrixToObjectList(arrays, cols);
+
+                        Array.prototype.push.apply(fullObjectList, objList);
+
+                        // convert the trace data from list to object and sort (backwards, stable sort)
+                        var sortedObjectList = sortLib.sortObjectList(cols, objList);
+                        var matrix = sortLib.objectListToList(sortedObjectList);
+                        var sortedMatrix = sortLib.sortedMatrix(matrix);
+
+                        axLabels = sortedMatrix[0].slice();
+                        var axVals = sortedMatrix[1];
+
+                        if(valLetter === 'z' & axLetter === 'x') {
+                            axVals = sortLib.transpose(axVals);
+                        }
+
+                        if(trace.type === 'ohlc' | trace.type === 'candlestick') {
+                            var sortedValsTransform = sortLib.transpose(axVals);
+                            gd._fullData[i].open = sortedValsTransform[0];
```

Comment:
Similar to my previous comment about adding things to `ax`, we can't be overwriting contents of `gd._fullData` in a way that conflicts with the contents of `gd.data`, but if necessary we can add underscore-prefixed attributes.

Also: this section looks a bit too tightly coupled with concerns that belong to individual trace types. If we need to have logic here that's specific to one or two trace types, that functionality should go in the trace module and get called from there, something like:
```js
var traceSetup = gd._fullData[i]._module.setupMultiCategory;
if(traceSetup) {
    traceSetup(gd._fullData[i], sortedValsTransform);
}
```

Then `candlestick` and `ohlc` would opt into this behavior by defining a `setupMultiCategory` function in their [index files](https://github.com/plotly/plotly.js/blob/master/src/traces/candlestick/index.js)